### PR TITLE
Query Builder

### PR DESCRIPTION
**Pull request by dantleech**
As PHPCR-ODM 1.3 has now been released we can now implement the QueryBuilder for the Document Manager.

For example:

``` php
        $builder
            ->setLocale('de')
            ->from()->document('page', 'p')->end()
            ->useStructure('p', 'overview')
            ->where()->eq()->structureField('p.article')->literal('hello');

        $query = $builder->getQuery();
```

### CODE REVIEW

File: src/Sulu/Component/Content/Document/Query/QueryBuilder.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+     *
+     * @throws InvalidArgumentException If a document with the given alias is already registered.
+     *
+     * @return $this
+     */
+    public function useStructure($documentAlias, $structureName)
+    {
+        if (isset($this->structureMap[$documentAlias])) {
+            throw new \InvalidArgumentException(sprintf(
+                'Structure "%s" is already registered for document alias "%s". Trying to register structure "%s"',
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/QueryBuilder.php]: Add test

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        MetadataFactoryInterface $metadataFactory,
+        PropertyEncoder $encoder,
+        DocumentStrategyInterface $strategy,
+        StructureMetadataFactory $structureMetaFactory
+    ) {
+        parent::__construct($session, $eventDispatcher, $metadataFactory, $encoder, $strategy);
+        $this->structureMetaFactory = $structureMetaFactory;
+    }
+
+    public function getQuery(BaseQueryBuilder $queryBuilder)
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: Add comment about this (`{@inheritdoc}`), and why the below type check is necessary

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getPhpcrProperty($alias, $field)
+    {
+        if (false !== $position = strpos($field, 'structure#')) {
+            $propertyName = substr($field, $position + 10);
+
+            if (!isset($this->structureMap[$alias])) {
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: Test

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+                    'No structure has been registered for document alias "%s". Use ->useStructure(\'%s\', \'structure_name\') to register a structure name. ' .
+                    'Registered document aliases: "%s"',
+                    $alias,
+                    $alias,
+                    implode('", "', array_keys($this->documentMetadata))
+                ));
+            }
+
+            if (!isset($this->documentMetadata[$alias])) {
+                throw new \InvalidArgumentException(sprintf(
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: Test

File: tests/Sulu/Component/Content/Document/Query/QueryBuilderTest.php [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<?php
+
```
Comment (dantleech) [inline tests/Sulu/Component/Content/Document/Query/QueryBuilderTest.php]: Add header

**Comment by danrot** [comment]
@dantleech Have you fixed your own comments here?

**Comment by dantleech** [comment]
I am not sure there current form is the best API, what about the following?

## Current
```php
        $builder
            ->setLocale('de')
            ->useStructure('p', 'overview')
            ->from()->document('page', 'p')->end()
            ->where()->eq()->field('p.structure#article')->literal('hello');
```

Proposed, use `structure` instead of `useStructure` and `structureField('p.article')` instead of `field('p.structure#article'):

```php
        $builder->structure('p', 'overview');
        $builder->where()->eq()->structureField('p.article')->literal('hello');
```

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php (hunk 4 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: This will be possibly redundant if we implement `structureField`

**Comment by dantleech** [comment]
This now supports `->structureField('p.article')` instead of `field('p.structure#article')`.

Note: there is a current issue with `orWhere()` and `andWhere()` not sure if this is an issue with the PHPCR-ODM iteself.

File: src/Sulu/Component/Content/Document/Query/Builder/ConstraintFactory.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    {
+        return $this->addChild(new ConstraintAndX($this));
+    }
+
+    public function not()
+    {
+        return $this->addChild(new ConstraintNot($this));
+    }
+
+    public function orX()
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/Builder/ConstraintFactory.php]: inherit docs

File: src/Sulu/Component/Content/Document/Query/Builder/ConstraintNot.php [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+class ConstraintNot extends ConstraintFactory implements SuluNodeInterface
+{
+    public function getCardinalityMap()
+    {
+        return array(
+            self::NT_CONSTRAINT => array(1, null),
+        );
+    }
+
+    public function getNodeType()
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/Builder/ConstraintNot.php]: inherit docs

File: src/Sulu/Component/Content/Document/Query/Builder/ConstraintOrX.php [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+class ConstraintOrX extends ConstraintFactory implements SuluNodeInterface
+{
+    public function getCardinalityMap()
+    {
+        return array(
+            self::NT_CONSTRAINT => array(1, null),
+        );
+    }
+
+    public function getNodeType()
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/Builder/ConstraintOrX.php]: inherit docs

File: src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php [hunk 9]
```diff
@@ -0,0 +1,19 @@
+<?php
+
+namespace Sulu\Component\Content\Document\Query;
+
+use Sulu\Component\Content\Document\Query\SuluNodeInterface;
+use Doctrine\ODM\PHPCR\Query\Builder\AbstractLeafNode;
+use Doctrine\ODM\PHPCR\Query\Builder\AbstractNode;
+
+class QueryBuilderUtil
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php]: Explanation.

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+use Sulu\Component\DocumentManager\DocumentStrategyInterface;
+use Sulu\Component\DocumentManager\MetadataFactoryInterface;
+use Sulu\Component\DocumentManager\Query\QueryBuilderConverter;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;
+use Doctrine\ODM\PHPCR\Query\Builder\AbstractNode;
+
+use Sulu\Component\Content\Document\Query\Builder as Sulu;
+use Doctrine\ODM\PHPCR\Query\Builder as PhpcrOdm;
+
+class StructureQueryBuilderConverter extends QueryBuilderConverter
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: Docs

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+{
+    private $structureMetaFactory;
+    private $structureMap = [];
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: Dics

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+                get_class($queryBuilder)
+            ));
+        }
+
+        $this->structureMap = $queryBuilder->getStructureMap();
+
+        return parent::getQuery($queryBuilder);
+    }
+
+    public function dispatch(AbstractNode $node)
```
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: inheritdocs

File: tests/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverterTest.php [hunk 13]
```diff
@@ -0,0 +1,45 @@
+<?php
+
+namespace Sulu\Component\Content\Document\Query;
+
+use Sulu\Component\Content\Document\Query\StructureQueryBuilderConverter;
+
+class StructureQueryBuilderConverterTest extends \PHPUnit_Framework_TestCase
```
Comment (dantleech) [inline tests/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverterTest.php]: Add note explaining that the converter is primarily tested in the functional test.

**Comment by dantleech** [comment]
orWhere problem seems to be a Sulu issue rather than a PHPCR-ODM one.

**Comment by dantleech** [comment]
@danrot @chirimoya @wachterjohannes ready for review.

**Comment by dantleech** [comment]
```
PHP Fatal error:  Class 'Sulu\Component\DocumentManager\Query\QueryBuilderConverter' not found in /home/travis/build/sulu-io/sulu/src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php on line 33
```

Guess we need to merge the DocumentManager PR or update `composer.json` here.

**Comment by dantleech** [comment]
Added `dev-query_builder_final` as DocumentManager version.

**Comment by danrot** [comment]
@dantleech Would have liked to review, but some of the tests are failing :confused:

File: composer.json [hunk 14]
```diff
@@ -24,7 +24,7 @@
         "imagine/imagine": "~0.6.1",
         "liip/theme-bundle": "1.3.*",
         "massive/search-bundle": "dev-develop",
-        "sulu/document-manager": "dev-develop",
+        "sulu/document-manager": "dev-query_builder_final",
```
Comment (danrot) [inline composer.json]: This is probably wrong now, you've renamed the branch to something using 2016, didn't you?

File: src/Sulu/Component/Content/Document/Query/Builder/ConstraintComparison.php [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+namespace Sulu\Component\Content\Document\Query\Builder;
+
+use Doctrine\ODM\PHPCR\Query\Builder\AbstractNode;
+use Doctrine\ODM\PHPCR\Query\Builder as Doctrine;
+use Sulu\Component\Content\Document\Query\QueryBuilderUtil;
+use Sulu\Component\Content\Document\Query\SuluNodeInterface;
+
+/**
+ * @see Doctrine\ODM\PHPCR\Query\Builder\ConstraintAndx
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/Builder/ConstraintComparison.php]: Wrong comment

File: src/Sulu/Component/Content/Document/Query/Builder/ConstraintFactory.php [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Content\Document\Query\Builder;
+
+use Doctrine\ODM\PHPCR\Query\Builder\ConstraintFactory as BaseConstraintFactory;
+use PHPCR\Query\QOM\QueryObjectModelConstantsInterface as QOMConstants;
+
+class ConstraintFactory extends BaseConstraintFactory
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/Builder/ConstraintFactory.php]: PHPdoc missing

File: src/Sulu/Component/Content/Document/Query/Builder/OrderByAdd.php [hunk 17]
```diff
@@ -0,0 +1,19 @@
...
+ */
+
+namespace Sulu\Component\Content\Document\Query\Builder;
+
+/**
+ * @see Doctrine\ODM\PHPCR\Query\Builder\OrderByAdd
+ */
+class OrderByAdd extends OrderBy
+{
+}
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/Builder/OrderByAdd.php]: Does this class serve a purpose?

File: src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ * (c) MASSIVE ART WebServices GmbH
+ *
+ * This source file is subject to the MIT license that is bundled
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Content\Document\Query;
+
+
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php]: only one new line

File: src/Sulu/Component/Content/Document/Query/Builder/OrderByAdd.php (hunk 17 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/Builder/OrderByAdd.php]: Yes. These are necessary, you will find an explanation in a comment in
the code.

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 19]
```diff
@@ -0,0 +1,222 @@
...
+        );
+
+        if (!method_exists($this, $methodName)) {
+            throw new \InvalidArgumentException(sprintf(
+                'Do not know how to walk node of type "%s"',
+                $node->getName()
+            ));
+        }
+
+        $res = $this->$methodName($node);
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: do not abbreviate

**Comment by danrot** [comment]
Same as in https://github.com/sulu-io/sulu-document-manager/pull/60, need to have a deeper look at it. If you can fix all my comments I will have this deeper look next week.

**Comment by dantleech** [comment]
Updated.

**Comment by danrot** [comment]
Tests are located in the old directory.

File: tests/Sulu/Component/Content/Document/Query/QueryBuilderFunctionalTest.php [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+ * We compare the query generated by the PHPCR layer against an expected query.
+ *
+ * WARNING: The query generated will not always match to the query we construct as the field names will
+ *          be translated to property names.
+ */
+class QueryBuilderFunctionalTest extends SuluTestCase
+{
+    public function setUp()
+    {
+        $this->manager = $this->getContainer()->get('sulu_document_manager.document_manager');
```
Comment (danrot) [inline tests/Sulu/Component/Content/Document/Query/QueryBuilderFunctionalTest.php]: Declaration missing

File: src/Sulu/Component/Content/Document/Query/SuluNodeInterface.php [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * All Sulu query builder nodes implement this interface.
+ */
+interface SuluNodeInterface
+{
+    /**
+     * Return a the Doctrine PHPCR-ODM instance of the node.
+     *
+     * See the documentation in the Sulu\Component\Content\Document\Query\StructureQueryBuilderConverter
+     * for information on why this is necessary.
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/SuluNodeInterface.php]: I would use the `@see` annotation here.

Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/SuluNodeInterface.php]: And this is necessary for auto completion, right? I think this comment is somewhere else, I can't really find that in the class mentioned here.

File: src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php [hunk 22]
```diff
@@ -0,0 +1,40 @@
...
+     * able to accept Sulu classes.
+     *
+     * @param SuluNodeInterface $suluNode
+     * @param AbstractNode $doctrineNode
+     *
+     * @return AbstractNode
+     */
+    public static function addNodeChildren(SuluNodeInterface $suluNode, AbstractNode $doctrineNode)
+    {
+        foreach ($suluNode->getChildren() as $childNode) {
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php]: The `getChildren` method is not known, because it's not in the `SuluNodeInterface`. Should the `SuluNodeInterface` maybe inherit form another interface?

Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php]: And thinking of it, should it have the part `Sulu` in its name? That's already part of the namespace.

File: src/Sulu/Component/Content/Document/Query/QueryBuilder.php [hunk 23]
```diff
@@ -0,0 +1,130 @@
...
+     * ````
+     * $qb->from()->document('foo', 'p');
+     * $qb->useStructure('p', 'overview');
+     * $qb->where()->eq()->structureField('p.field1')->literal('barbar');
+     * ````
+     *
+     * @param mixed $documentAlias Document alias to register structure to
+     * @param mixed $structureName Name of the structure to register.
+     *
+     * @throws InvalidArgumentException If a document with the given alias is already registered.
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/QueryBuilder.php]: use missing

File: src/Sulu/Component/Content/Document/Query/Builder/OperandFactory.php [hunk 24]
```diff
@@ -0,0 +1,43 @@
...
+     *     ->literal('my_field_value')
+     *   ->end()
+     * ->end();
+     * </code>
+     *
+     * @param string $field - name of field to check, including alias name.
+     *
+     * @factoryMethod OperandDynamicField
+     *
+     * @return OperandDynamicFactory
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/Builder/OperandFactory.php]: use missing

File: src/Sulu/Component/Content/Document/Query/Builder/ConstraintComparison.php [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+    {
+        return $this->operator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getDoctrineInstance()
+    {
+        return QueryBuilderUtil::addNodeChildren($this, new Doctrine\ConstraintComparison($this->getParent(), $this->getOperator()));
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/Builder/ConstraintComparison.php]: Line too long

File: src/Sulu/Component/Content/Document/Query/SuluNodeInterface.php (hunk 21 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/SuluNodeInterface.php]: No, it is necessary because the PHPCR-ODM query converter will not accept our node classes. The documentation your referring to is in `QueryBuilder` class.

Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/SuluNodeInterface.php]: (I added a note here).

File: src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php (hunk 22 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php]: There are two nodes involved, one is from Sulu the other from Doctrine.

Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/QueryBuilderUtil.php]: Removed the SuluNode type hint as it works with any `AbstractNode`. The PHPCR-ODM has no interface that the SuluNodeInterface can extend.

**Comment by dantleech** [comment]
Updated and moved tests

File: src/Sulu/Component/Content/Document/Query/QueryBuilder.php [hunk 26]
```diff
@@ -0,0 +1,130 @@
...
+     * ````
+     *
+     * @param string $documentAlias Document alias to register structure to
+     * @param string $structureName Name of the structure to register.
+     *
+     * @throws \InvalidArgumentException If a document with the given alias is already registered.
+     *
+     * @return self
+     */
+    public function useStructure($documentAlias, $structureName)
```
Comment (webmozart) [inline src/Sulu/Component/Content/Document/Query/QueryBuilder.php]: Shouldn't you invert the order of these arguments? That would be consistent with Doctrine ORM (e.g. `innerJoin('p.article', 'a')`).

Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/QueryBuilder.php]: .. yeah we could do that. That would also allow us to a default value
for document alias when there is only one document.

File: src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php [hunk 27]
```diff
@@ -0,0 +1,220 @@
...
+    {
+        return parent::walkConstraintComparison($node->getDoctrineInstance());
+    }
+
+    /**
+     * Walk the Sulu StructureField contsraint.
+     *
+     * @param Sulu\OperandDynamicStructureField $node
+     *
+     * @return PHPCR\Query\QOM\PropertyValue
```
Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: Wrong FQN, leading `\` missing.

Comment (danrot) [inline src/Sulu/Component/Content/Document/Query/StructureQueryBuilderConverter.php]: Actually this class does not exist at all.

File: src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php [hunk 28]
```diff
@@ -0,0 +1,260 @@
...
+ *
+ * We compare the query generated by the PHPCR layer against an expected query.
+ *
+ * WARNING: The query generated will not always match to the query we construct as the field names will
+ *          be translated to property names.
+ */
+class QueryBuilderFunctionalTest extends SuluTestCase
+{
+    /**
+     * @var DocumentManager
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php]: Usage missing

File: src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php [hunk 29]
```diff
@@ -0,0 +1,260 @@
...
+        $this->assertQuery(
+            'SELECT * FROM [nt:unstructured] AS p WHERE p.[jcr:mixinTypes] = \'sulu:page\' ORDER BY p.[i18n:de-nodeType] ASC, p.[i18n:de-article] DESC',
+            $builder
+        );
+    }
+
+    /**
+     * It should throw an exception if an unknown structure field is used.
+     *
+     * @expectedException InvalidArgumentException
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php]: Leading `/` missing.

File: src/Sulu/Component/Content/Tests/Unit/Document/Query/QueryBuilderTest.php [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+ * with this source code in the file LICENSE.
+ */
+
+namespace Sulu\Component\Content\Tests\Unit\Document\Query;
+
+use Sulu\Component\Content\Document\Query\QueryBuilder;
+
+class QueryBuilderTest extends \PHPUnit_Framework_TestCase
+{
+    private $queryBuilder;
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Unit/Document/Query/QueryBuilderTest.php]: PHPDoc missing

File: src/Sulu/Component/Content/Tests/Unit/Document/Query/QueryBuilderTest.php [hunk 31]
```diff
@@ -0,0 +1,48 @@
...
+        $this->queryBuilder->useStructure('p', 'overview');
+        $this->assertEquals([
+            'p' => 'overview',
+        ], $this->queryBuilder->getStructureMap());
+    }
+
+    /**
+     * It should throw an exception if the structure has already been assigned.
+     *
+     * @expectedException InvalidArgumentException
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Unit/Document/Query/QueryBuilderTest.php]: Leading `\` missing.

File: src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php [hunk 32]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+namespace Sulu\Component\Tests\Unit\Content\Document\Query;
+
+use Sulu\Component\Content\Document\Query\StructureQueryBuilderConverter;
+
+class StructureQueryBuilderConverterTest extends \PHPUnit_Framework_TestCase
+{
+    private $metadataFactory;
+    private $encoder;
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php]: PHPDocs missing

File: src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php [hunk 33]
```diff
@@ -0,0 +1,54 @@
...
+        $this->encoder = $this->prophesize('Sulu\Bundle\DocumentManagerBundle\Bridge\PropertyEncoder');
+        $session = $this->prophesize('PHPCR\SessionInterface');
+        $workspace = $this->prophesize('PHPCR\WorkspaceInterface');
+        $queryManager = $this->prophesize('PHPCR\Query\QueryManagerInterface');
+        $qomf = $this->prophesize('PHPCR\Query\QOM\QueryObjectModelFactoryInterface');
+        $session->getWorkspace()->willReturn($workspace->reveal());
+        $workspace->getQueryManager()->willReturn($queryManager->reveal());
+        $queryManager->getQOMFactory()->willReturn($qomf->reveal());
+
+        $this->converter = new StructureQueryBuilderConverter(
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php]: Not declared

File: src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php [hunk 34]
```diff
@@ -0,0 +1,54 @@
...
+    private $encoder;
+
+    public function setUp()
+    {
+        $this->metadataFactory = $this->prophesize('Sulu\Component\Content\Metadata\Factory\StructureMetadataFactory');
+        $this->encoder = $this->prophesize('Sulu\Bundle\DocumentManagerBundle\Bridge\PropertyEncoder');
+        $session = $this->prophesize('PHPCR\SessionInterface');
+        $workspace = $this->prophesize('PHPCR\WorkspaceInterface');
+        $queryManager = $this->prophesize('PHPCR\Query\QueryManagerInterface');
+        $qomf = $this->prophesize('PHPCR\Query\QOM\QueryObjectModelFactoryInterface');
```
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php]: Not declared

File: src/Sulu/Component/Content/Document/Query/QueryBuilder.php (hunk 26 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Document/Query/QueryBuilder.php]: have reversed and the `documentAlias` is optional and can be omitted if only one Document is invovled in the query.

File: src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php (hunk 29 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php]: Nope. These annotations are always resolved from the base namespace.

File: src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php (hunk 34 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php]: What is not declared?

**Comment by dantleech** [comment]
Updated.

File: src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php (hunk 29 above)
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php]: Well, IntelliJ marks that as wrong, and I think PHPUnit does not care.

File: src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php (hunk 33 above)
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Unit/Document/Query/StructureQueryBuilderConverterTest.php]: This variable is still not declared :wink:

File: src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php (hunk 29 above)
Comment (danrot) [inline src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php]: Btw, I think it's marked as wrong because it might be ambiguous. If there is another `InvalidArgumentException` added to this namespace later, this might lead to wrong results.

Comment (dantleech) [inline src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php]: PHPUnit will always evaluate this absolutely. It is not aware of
namespaces here.

Comment (danrot) [inline src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php]: As said, it's not about PHPUnit here. It's marked as an mistake in IntelliJ, and I would like to have as less of them as possible.

**Comment by danrot** [comment]
We are also updating to phpcr-odm 1.3 here, right? Then we could probably remove these lines: https://github.com/sulu-io/sulu/blob/7bbdcd18d50a320ad2c7e6286138a1b8a11d7901/src/Sulu/Bundle/CoreBundle/DependencyInjection/SuluCoreExtension.php#L44

[The linked issue](https://github.com/doctrine/DoctrinePHPCRBundle/issues/178) is already closed.

File: src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php (hunk 29 above)
Comment (dantleech) [inline src/Sulu/Component/Content/Tests/Functional/Document/Query/QueryBuilderFunctionalTest.php]: Fixed this instance, but this is my default behavior, so there are bound
to be many other examples which should be fixed in a separate PR.

**Comment by dantleech** [comment]
@danrot lets not remove those lines in this PR. It is a separate issue and I don't know what side effects that will have or if it will be picked up by the tests.

**Comment by webmozart** [comment]
@dantleech Could you give a more elaborate example of how this would be used?

**Comment by dantleech** [comment]
@webmozart it is based upon the PHPCR-ODM query builder, every thing [here](http://doctrine-orm.readthedocs.org/projects/doctrine-phpcr-odm/en/latest/reference/query-builder.html) applies to this query builder, we add the `useStructure` and `structureField` methods allowing you to query based on the "Structure" fields - the things that hold the actual user content in Sulu.

**Comment by webmozart** [comment]
@danrot thanks!
